# EDA - Fuente de datos 2
## Plan Anual de Adquisiones - Secop II

**Distribución de asignaciones:

**Daniela:** Ajuste de NIT, Entidades y contactos - Verificar cuales son las identidades repetidas, ajustarlo y cargar nit en fuente 3

**Daniel:** Localización y descripción - Cruzar Divipola y cargar Geolocalizacion

**Yaliza:** Misión/Vision, Perspectivas - Tokenizacion, analisis del lenguaje, busqueda de transformacion digital  - Revisión de categoria de producto y variable "version"

**David:** Ajuste de fecha, evaluación del mes proyectado, decidir fecha de publicación

**Cristian:** Revisión de variables monetarias, ajuste de presupuesto y revision de precios bases en cero y variables "Referencia"

**Importar librerías.**

In [2]:
# Load Packages
import os,sys
import pandas as pd
import numpy as np
from datetime import datetime

# This line is needed to display plots inline in Jupyter Notebook
%matplotlib inline

# Required for basic python plotting functionality
import matplotlib.pyplot as plt

# Required for formatting dates later in the case
import datetime
import matplotlib.dates as mdates

# Required to display image inline
from IPython.display import Image

# Advanced plotting functionality with seaborn
import seaborn as sns
sns.set(style="whitegrid") # can set style depending on how you'd like it to look

**Importar la base de datos (CSV).**

In [3]:
df = pd.read_csv ('Plan_Anual_de_Adquisiciones_-_SECOP_II.csv')
df.head()

Anno Identificador PAA                                            Entidad  \
0  2018    CO1.APP.361029  COMANDO DE RECLUTAMIENTO Y CONTROL RESERVAS DE...   
1  2019    CO1.APP.525808                     SUPERINTENDENCIA DE TRANSPORTE   
2  2019    CO1.APP.460110                                       FONAM - DTAN   
3  2018    CO1.APP.338501                           ESAP TERRITORIAL BOLIVAR   
4  2018    CO1.APP.324303                             GRUPO AÉREO DEL CARIBE   

         NIT  Localización                      DescripcionUbicacion  \
0  800130643   CO-DC-11001                Distrito Capital de Bogotá   
1  800170433   CO-DC-11001                Distrito Capital de Bogotá   
2  901037393  CO-SAN-68001                                 Santander   
3  800248529  CO-BOL-13001                                   Bolívar   
4  800141634  CO-SAP-88001  San Andrés, Providencia y Santa Catalina   

                                       Mision/Vision  \
0  MISIÓN: El Comando De Reclutamiento y Control ...   
1  Vigilar, inspeccionar y controlar el cumplimie...   
2  Misión: Conservamos las áreas protegidas del S...   
3  Formar ciudadanos y ciudadanas en los conocimi...   
4  MISIÓN: Ejecutar operaciones aéreas para contr...   

                             Perspectiva Estrategica  \
0  Con el fin de cumplir con la misión institucio...   
1  La Superintendencia de Puertos y Transporte vi...   
2  La entidad estatal va a contratar los procesos...   
3  La ESAP en el 2019 será una institución de car...   
4  Consolidar una Fuerza Decisiva y efectiva en c...   

   Presupuesto Menor Cuantia  Presupuesto Minima Cuantia  ...  \
0                781242000.0                  78124200.0  ...   
1                231872480.0                  23187248.0  ...   
2                372652210.0                  37265221.0  ...   
3                507807300.0                  50780730.0  ...   
4                781242000.0                  78124200.0  ...   

        Referencia Operacion Fecha Publicacion  \
0             131-DIREC-2018           36:51.8   
1              IPMC 009-2019           59:02.7   
2           DTAN-CD-032-2019           05:15.4   
3             CD-07-092-2018           46:31.2   
4  010-00-N-GACAR-ESDEB-2018           28:01.6   

                                Modalidad                         Contacto  \
0  CCE-16-Servicios_profesionales_gestion      ELKIN ALONSO ARGOTE HIDALGO   
1                   CCE-10-Minima_Cuantia     MARIA PIERINA GONZALEZ FALLA   
2  CCE-16-Servicios_profesionales_gestion         RUTH MILADY MEJIA OLARTE   
3  CCE-16-Servicios_profesionales_gestion                  KAROLAIN MORENO   
4                   CCE-10-Minima_Cuantia  Diana katherine Quintana Cuervo   

  UNSPSC - Codigo Producto                      UNSPSC - Nombre Producto  \
0              V1.80121700            Servicios de responsabilidad civil   
1              V1.53101900                                        Trajes   
2              V1.80111701         Servicios de contratación de personal   
3              V1.80111600                Servicios de personal temporal   
4              V1.10111302  Productos para el aseo y cuidado de mascotas   

  UNSPSC - Codigo Clase                              UNSPSC - Nombre Clase  \
0           V1.80120000                                  Servicios legales   
1           V1.53100000                                               Ropa   
2           V1.80111700                          Reclutamiento de personal   
3           V1.80110000                      Servicios de recursos humanos   
4           V1.10111300  Tratamientos para los animales domésticos y ac...   

  UNSPSC - Codigo Familia                            UNSPSC - Nombre Familia  
0             V1.80000000  Servicios de Gestión, Servicios Profesionales ...  
1             V1.53000000         Ropa, Maletas y Productos de Aseo Personal  
2             V1.80110000                      Servicios de recursos humanos  
3           

**Tamaño de la base de datos, nombres de variables y tipo de datos.**

In [4]:
df.shape

(863969, 29)

In [5]:
df.dtypes

Anno                            int64
Identificador PAA              object
Entidad                        object
NIT                             int64
Localización                   object
DescripcionUbicacion           object
Mision/Vision                  object
Perspectiva Estrategica        object
Presupuesto Menor Cuantia     float64
Presupuesto Minima Cuantia    float64
Presupuesto Global            float64
Fecha Primera Publicación      object
Mes Proyectado                 object
Identificador Item             object
Categoria Principal            object
Precio Base                   float64
Ultima Fecha Modificacion      object
Version                         int64
Referencia Contrato            object
Referencia Operacion           object
Fecha Publicacion              object
Modalidad                      object
Contacto                       object
UNSPSC - Codigo Producto       object
UNSPSC - Nombre Producto       object
UNSPSC - Codigo Clase          object
UNSPSC - Nom

## Organizar los datos para una vista minable.

In [6]:
# Renombrar las columnas 
df  = df.rename(columns={
     'Anno':                        "year", 
     'Identificador PAA':           "id_paa",
     'Entidad':                     "entidad",
     'NIT':                         "nit",
     'Localización':                "localizacion",
     'DescripcionUbicacion':        "localizacion_desc",
     'Mision/Vision':               "mision_vision",
     'Perspectiva Estrategica':     "pers_estrategica",
     'Presupuesto Menor Cuantia':   "ppto_menor_cuantia",
     'Presupuesto Minima Cuantia':  "ppto_min_cuantia",
     'Presupuesto Global':          "ppto_global",
     'Fecha Primera Publicación':   "date_first_publication",
     'Mes Proyectado':              "mes_proyectado",
     'Identificador Item':          "id_item",
     'Categoria Principal':         "categoria_principal",
     'Precio Base':                 "precio_base",
     'Ultima Fecha Modificacion':   "date_last_publication",
     'Version':                     "version",
     'Referencia Contrato':         "ref_contrato",
     'Referencia Operacion':        "ref_operacion",
     'Fecha Publicacion':           "date_publised",
     'Modalidad':                   "modalidad",
     'Contacto':                    "contacto",
     'UNSPSC - Codigo Producto':    "cod_producto", 
     'UNSPSC - Nombre Producto':    "nombre_producto", 
     'UNSPSC - Codigo Clase':       "cod_clase",  
     'UNSPSC - Nombre Clase':       "nombre_clase", 
     'UNSPSC - Codigo Familia':     "cod_familia", 
     'UNSPSC - Nombre Familia':     "nombre_familia"
})

In [7]:
print(list(df.columns))

['year', 'id_paa', 'entidad', 'nit', 'localizacion', 'localizacion_desc', 'mision_vision', 'pers_estrategica', 'ppto_menor_cuantia', 'ppto_min_cuantia', 'ppto_global', 'date_first_publication', 'mes_proyectado', 'id_item', 'categoria_principal', 'precio_base', 'date_last_publication', 'version', 'ref_contrato', 'ref_operacion', 'date_publised', 'modalidad', 'contacto', 'cod_producto', 'nombre_producto', 'cod_clase', 'nombre_clase', 'cod_familia', 'nombre_familia']


In [8]:
# Validar los valores únicos de las variables
df.nunique(axis=0)
print(df.nunique(axis=0))

year                           5
id_paa                      2091
entidad                     1057
nit                          941
localizacion                 316
localizacion_desc             33
mision_vision               1648
pers_estrategica            1523
ppto_menor_cuantia           241
ppto_min_cuantia             211
ppto_global                 2086
date_first_publication      2029
mes_proyectado                13
id_item                   241220
categoria_principal         6192
precio_base                71796
date_last_publication        827
version                      174
ref_contrato              197879
ref_operacion             197887
date_publised               2029
modalidad                     13
contacto                    1643
cod_producto                6192
nombre_producto             6177
cod_clase                   1856
nombre_clase                1852
cod_familia                  414
nombre_familia               413
dtype: int64


Tenemos 941 NIT únicos con 1057 nombres de entidades.                          

**¿Cómo identificar los 116 nombres que tienen un NIT ya utilizado?**

In [9]:
#Generar tabla dinamica, dejando NIT como valor del index

df_pivot = pd.pivot_table(df,index=["entidad"],values=["nit"])
print(df_pivot)
df_pivot.dtypes

                                                           nit
entidad                                                       
ADMINISTRADORA DE LOS RECURSOS DEL SISTEMA GENE...   901037916
AGENCIA DE DESARROLLO RURAL                          900948958
AGENCIA DE EDUCACIÓN POSTSECUNDARIA DE MEDELLÍN...   900602106
AGENCIA DE RENOVACION DEL TERRITORIO                 901006886
AGENCIA LOGISTICA DE LAS FUERZAS MILITARES           899999162
...                                                        ...
UNIDAD PRESTADORA DE SALUD CÓRDOBA                   900800975
UNIDAD PRESTADORA DE SALUD NARIÑO                    900807338
UNIDAD PRESTADORA DE SALUD NORTE DE SANTANDER -...   900259415
UNIVERSIDAD DE CUNDINAMARCA                         8906800622
VEEDURIA DISTRITAL                                   899999061

[1057 rows x 1 columns]


nit    int64
dtype: object

In [10]:
# Select all duplicate rows based on one column
duplicateRowsDF = df_pivot[df_pivot.duplicated(['nit'])]
print("Duplicate Rows based on a single column are:", duplicateRowsDF, sep='\n')

Duplicate Rows based on a single column are:
                                                          nit
entidad                                                      
ALCALDIA LOCAL DE CIUDAD BOLIVAR                    899999061
ALCALDIA LOCAL DE ENGATIVÁ                          899999061
ALCALDIA LOCAL DE FONTIBON                          899999061
ALCALDIA LOCAL DE KENNEDY                           899999061
ALCALDIA LOCAL DE MARTIRES                          899999061
...                                                       ...
UNIDAD PRESTADORA DE SALUD ARAUCA - DEPARTAMENT...  800140602
UNIDAD PRESTADORA DE SALUD CÓRDOBA                  900800975
UNIDAD PRESTADORA DE SALUD NARIÑO                   900807338
UNIDAD PRESTADORA DE SALUD NORTE DE SANTANDER -...  900259415
VEEDURIA DISTRITAL                                  899999061

[116 rows x 1 columns]


##### Creamos un indice de NIT para asignar el nombre de la entidad matriz a cada registro

In [11]:
indice_nit = pd.read_csv('index_nit.csv', encoding= 'unicode_escape')
indice_nit.head()

NIT                                 entidad
0   40330032        BARBARA IVONNE LINARES RODRIGUEZ
1   88143760            SAID ALONSO PENALOZA ANTELIZ
2  123456780                             NO UTILIZAR
3  800005042  CONTRALORIA MUNICIPAL DE FLORIDABLANCA
4  800006850    HOSPITAL MARIO GAITAN YANGUAS E.S.E

Aseguramos que la variable "NIT" conserve el mismo formato en ambos lados y modificamos el nombre de entidad por "entidad_matriz"

In [12]:
#Renombrar variables
indice_nit  = indice_nit.rename(columns={'NIT': "nit",
                                        'entidad': "entidad_matriz"}) 
indice_nit

nit                                     entidad_matriz
0      40330032                   BARBARA IVONNE LINARES RODRIGUEZ
1      88143760                       SAID ALONSO PENALOZA ANTELIZ
2     123456780                                        NO UTILIZAR
3     800005042             CONTRALORIA MUNICIPAL DE FLORIDABLANCA
4     800006850               HOSPITAL MARIO GAITAN YANGUAS E.S.E
..          ...                                                ...
936  9008012094                      POLICIAMETROPOLITANA DE TUNJA
937  9008529985                 AGENCIA NACIONAL DE SEGURIDAD VIAL
938  9009585649   SUBRED INTEGRADA DE SERVICIOS DE SALUD SUR E.S.E
939  9011548528  SOCIEDAD PUBLICA TERMINAL REGIONAL DE TRANSPOR...
940  9013620904               UNIDAD PRESTADORA DE SALUD ATLANTICO

[941 rows x 2 columns]

Agregamos un merge entre la matriz de entidades y el dataframe.

In [13]:
from pandas import DataFrame, merge

In [14]:
df = indice_nit.merge(df, on='nit', how='left')
df

nit                                     entidad_matriz  year  \
0         40330032                   BARBARA IVONNE LINARES RODRIGUEZ  2019   
1         88143760                       SAID ALONSO PENALOZA ANTELIZ  2018   
2        123456780                                        NO UTILIZAR  2016   
3        800005042             CONTRALORIA MUNICIPAL DE FLORIDABLANCA  2017   
4        800006850               HOSPITAL MARIO GAITAN YANGUAS E.S.E  2019   
...            ...                                                ...   ...   
863964  9011548528  SOCIEDAD PUBLICA TERMINAL REGIONAL DE TRANSPOR...  2018   
863965  9011548528  SOCIEDAD PUBLICA TERMINAL REGIONAL DE TRANSPOR...  2018   
863966  9011548528  SOCIEDAD PUBLICA TERMINAL REGIONAL DE TRANSPOR...  2018   
863967  9013620904               UNIDAD PRESTADORA DE SALUD ATLANTICO  2020   
863968  9013620904               UNIDAD PRESTADORA DE SALUD ATLANTICO  2020   

                id_paa                                            entidad  \
0       CO1.APP.361052                   BARBARA IVONNE LINARES RODRIGUEZ   
1       CO1.APP.193629                       SAID ALONSO PEÑALOZA ANTELIZ   
2        CO1.APP.11902                                        NO UTILIZAR   
3       CO1.APP.155102             CONTRALORIA MUNICIPAL DE FLORIDABLANCA   
4       CO1.APP.372454                HOSPITAL MARIO GAITÁN YANGUAS E.S.E   
...                ...                                                ...   
863964  CO1.APP.246214  SOCIEDAD PUBLICA TERMINAL REGIONAL DE TRANSPOR...   
863965  CO1.APP.246214  SOCIEDAD PUBLICA TERMINAL REGIONAL DE TRANSPOR...   
863966  CO1.APP.246214  SOCIEDAD PUBLICA TERMINAL REGIONAL DE TRANSPOR...   
863967  CO1.APP.590905               UNIDAD PRESTADORA DE SALUD ATLANTICO   
863968  CO1.APP.590905               UNIDAD PRESTADORA DE SALUD ATLANTICO   

        localizacion           localizacion_desc  \
0             CO-MET                        Meta   
1       CO-NSA-54498          Norte de Santander   
2        CO-DC-11001  Distrito Capital de Bogotá   
3       CO-SAN-68276                   Santander   
4       CO-CUN-25754                Cundinamarca   
...              ...                         ...   
863964        CO-BOY                      Boyacá   
863965        CO-BOY                      Boyacá   
863966        CO-BOY                      Boyacá   
863967  CO-ATL-08758                   Atlántico   
863968  CO-ATL-08758                   Atlántico   

                                            mision_vision  \
0       El municipio de La Macarena será en el 2027 un...   
1       Promover el desarrollo humano y sostenible del...   
2       Misión\nLograr una EDUCACIÓN DE CALIDAD, que f...   
3       MISIÓN: Fortalecer el control y la vigilancia ...   
4       Misión\n\nSomos una institución pública presta...   
...                                                   ...   
863964  Promover, desarrollar y explotar soluciones de...   
863965  Promover, desarrollar y explotar soluciones de...   
863966  Promover, desarrollar y explotar soluciones de...   
863967  Contribuir a la calidad de vida de nuestros us...   
863968  Contribuir a la calidad de vida de nuestros us...   

                                         pers_estrategica  ppto_menor_cuantia  \
0       Un Gobierno que responda oportuna y eficientem...         231872480.0   
1       En el año 2025, El Municipio de Ocaña será una...         218747844.0   
2       Objetivos Misionales\n\n•Brindar educación ini...         689454000.0   
3       1.Fortalecer la vigilancia y el control fiscal...         206560760.0   
4       La ESE Ese Hospital Mario Gaitán del municipio...         165623200.0   
...                                                   ...                 ...   
863964  Asegurar que el sistema de compras y contratac...         218747760.0   
863965  Asegurar que el sistema de compras y contratac...         218747760.0   
863966  Asegurar que el sistema de compras y contratac

In [16]:
#Generar tabla dinamica, dejando NIT como valor del index

df_pivot = pd.pivot_table(df,index=["entidad_matriz","entidad"],values=["nit"])
print(df_pivot)
df_pivot.dtypes

                                                                                                              nit
entidad_matriz                                     entidad                                                       
ADMINISTRADORA DE LOS RECURSOS DEL SISTEMA GENE... ADMINISTRADORA DE LOS RECURSOS DEL SISTEMA GENE...   901037916
AGENCIA DE DESARROLLO RURAL                        AGENCIA DE DESARROLLO RURAL                          900948958
AGENCIA DE EDUCACIN POSTSECUNDARIA DE MEDELLN ... AGENCIA DE EDUCACIÓN POSTSECUNDARIA DE MEDELLÍN...   900602106
AGENCIA DE RENOVACION DEL TERRITORIO               AGENCIA DE RENOVACION DEL TERRITORIO                 901006886
AGENCIA LOGISTICA DE LAS FUERZAS MILITARES         AGENCIA LOGISTICA DE LAS FUERZAS MILITARES           899999162
...                                                                                                           ...
UNIDAD NACIONAL DE PROTECCIN                       UNIDAD NACIONAL DE PROTECCIÓN        

nit    int64
dtype: object

In [17]:
# Select all duplicate rows based on one column
duplicateRowsDF = df_pivot[df_pivot.duplicated(['nit'])]
print("Duplicate Rows based on a single column are:", duplicateRowsDF, sep='\n')

Duplicate Rows based on a single column are:
                                                                                                              nit
entidad_matriz                                     entidad                                                       
ALCALDIA MUNICIPAL DE IZA                          CONCEJO MUNICIPIO DE IZA                             891856077
BOGOTA DISTRITO CAPITAL                            ALCALDIA LOCAL DE CIUDAD BOLIVAR                     899999061
                                                   ALCALDIA LOCAL DE ENGATIVÁ                           899999061
                                                   ALCALDIA LOCAL DE FONTIBON                           899999061
                                                   ALCALDIA LOCAL DE KENNEDY                            899999061
...                                                                                                           ...
POLICIA NACIONA - DIRECCION ADMINISTRATIVA 

In [18]:
df_fuente3 = pd.read_csv('fuente_3.csv', encoding= 'unicode_escape')
df_fuente3.head()

nit         entidad  top_30_innovadoras_2019  \
0  890100251  Cementos_Argos                      1.0   
1  900696206          Corona                      2.0   
2  890106527         Procaps                      3.0   
3  899999068       Ecopetrol                      4.0   
4  900062201           Alsec                      5.0   

   top_10_con_mas_doctores_I+D  top_Ecosistemas_de_innovacion  top_Patentes  \
0                          2.0                            1.0           NaN   
1                          NaN                            2.0           5.0   
2                          6.0                            NaN           9.0   
3                          1.0                            NaN           4.0   
4                          NaN                            3.0           NaN   

   top_Nuevos_Ecosistemas  top_Nuevos_Productos  
0                     NaN                   NaN  
1                     NaN                   2.0  
2                     NaN                   NaN  
3                     NaN                   NaN  
4                     NaN                   NaN

In [20]:
df['fuente3']=df_fuente3.nit.isin(df.nit)
print(df['fuente3'].value_counts())

False    47
Name: fuente3, dtype: int64
